media/client/main/source/WebAudioPlayer.cpp

```diff
@@ -60,83 +67,163 @@ std::unique_ptr<IWebAudioPlayer> WebAudioPlayerFactory::createWebAudioPlayer(std
+        throw std::runtime_error("Web audio player ipc could not be created");
+    }
 }
 
 WebAudioPlayer::~WebAudioPlayer()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
+
+    m_webAudioPlayerIpc.reset();
+    if (!m_sharedMemoryManager->unregisterClient(this))
+    {
+        RIALTO_CLIENT_LOG_WARN("Failed to unregister client with SharedMemoryManager");
+    }
+
+    m_sharedMemoryManager.reset();
 }
 
 bool WebAudioPlayer::play()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->play();
 }
 
 bool WebAudioPlayer::pause()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->pause();
 }
 
 bool WebAudioPlayer::setEos()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->setEos();
 }
 
-bool WebAudioPlayer::getBufferAvailable(uint32_t &availableFrames, std::shared_ptr<WebAudioShmInfo> &webAudioShmInfo)
+bool WebAudioPlayer::getBufferAvailable(uint32_t &availableFrames, std::shared_ptr<WebAudioShmInfo> &)
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
-
-    return false;
+    if (!m_webAudioShmInfo)
+    {
+        m_webAudioShmInfo = std::make_shared<WebAudioShmInfo>();
+    }
+    std::lock_guard<std::mutex> bufLocker(m_bufLock);
+    return m_webAudioPlayerIpc->getBufferAvailable(availableFrames, m_webAudioShmInfo);
 }
 
 bool WebAudioPlayer::getBufferDelay(uint32_t &delayFrames)
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->getBufferDelay(delayFrames);
 }
 
 bool WebAudioPlayer::writeBuffer(const uint32_t numberOfFrames, void *data)
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    if (!m_webAudioShmInfo)
+    {
+        RIALTO_CLIENT_LOG_ERROR("Web audo shared info is null!");
+        return false;
+    }
+
+    uint32_t dataLength = numberOfFrames * 4;
+    uint32_t availableDataLength = m_webAudioShmInfo->lengthMain + m_webAudioShmInfo->lengthWrap;
+    if (dataLength > availableDataLength)
+    {
+        RIALTO_CLIENT_LOG_ERROR("The number of frames to write exceeds the available space!");
+        return false;
+    }
+
+    uint8_t *shmBuffer = m_sharedMemoryManager->getSharedMemoryBuffer();
+    if (nullptr == shmBuffer)
+    {
+        RIALTO_CLIENT_LOG_ERROR("Shared buffer no longer valid");
+        return false;
+    }
+
+    if (dataLength > m_webAudioShmInfo->lengthMain)
+    {
+        std::memcpy(shmBuffer + m_webAudioShmInfo->offsetMain, data, m_webAudioShmInfo->lengthMain);
+        std::memcpy(shmBuffer + m_webAudioShmInfo->offsetWrap,
+                    reinterpret_cast<uint8_t *>(data) + m_webAudioShmInfo->lengthMain,
+                    dataLength - m_webAudioShmInfo->lengthMain);
+    }
+    else
+    {
+        std::memcpy(shmBuffer + m_webAudioShmInfo->offsetMain, data, m_webAudioShmInfo->lengthMain);
+    }
+    std::lock_guard<std::mutex> bufLocker(m_bufLock);
```

Comment:
I think this lock needs to protect reading from m_webAudioShmInfo too, we could have a race condition where there is an ongoing getBufferAvailable which will change the m_webAudioShmInfo while we are writing data to the shmBuffer.

Reply:
@skyrangelovv, these are still not addressed if you weren't aware.